src/raw_node.rs

```diff
@@ -349,94 +385,299 @@ impl<T: Storage> RawNode<T> {
+        if rd_record.snap_meta.is_some() {
+            raft.raft_log.stable_snap();
+        }
+        if rd_record.last_entry.is_some() {
+            raft.raft_log.stable_entries();
+        }
     }
 
-    /// Grabs the snapshot from the raft if available.
-    #[inline]
-    pub fn snap(&self) -> Option<&Snapshot> {
-        self.raft.snap()
+    fn commit_apply(&mut self, applied: u64) {
+        self.raft.commit_apply(applied);
     }
 
-    /// Advance notifies the RawNode that the application has applied and saved progress in the
-    /// last Ready results.
-    pub fn advance(&mut self, rd: Ready) {
-        self.advance_append(rd);
-        let commit_idx = self.prev_hs.commit;
-        if commit_idx != 0 {
-            // In most cases, prevHardSt and rd.HardState will be the same
-            // because when there are new entries to apply we just sent a
-            // HardState with an updated Commit value. However, on initial
-            // startup the two are different because we don't send a HardState
-            // until something changes, but we do send any un-applied but
-            // committed entries (and previously-committed entries may be
-            // incorporated into the snapshot, even if rd.CommittedEntries is
-            // empty). Therefore we mark all committed entries as applied
-            // whether they were included in rd.HardState or not.
-            self.advance_apply(commit_idx);
+    /// Notifies that the ready of this number has been well persisted.
+    ///
+    /// Since Ready must be persisted in order, calling this function implicitly means
+    /// all readys with numbers smaller than this have been persisted.
+    pub fn on_persist_ready(&mut self, number: u64) {
+        while let Some(record) = self.records.front() {
+            if record.number > number {
+                break;
+            }
+            let mut record = self.records.pop_front().unwrap();
+
+            if let Some((index, term)) = record.snap_meta {
+                self.raft.on_persist_snap(index, term);
+                self.pending_snapshot = false;
+            }
+
+            if let Some((index, term)) = record.last_entry {
+                self.raft.on_persist_entries(index, term);
+            }
+
+            if !record.messages.is_empty() {
+                self.messages.push(mem::take(&mut record.messages));
+            }
+        }
+    }
+
+    /// Notifies that the last ready has been well persisted.
+    ///
+    /// Returns the PersistLastReadyResult that contains commit index, committed entries
+    /// and messages.
+    ///
+    /// This function must be called before entering the next round(e.g. propose, step).
+    /// Since Ready must be persisted in order, calling this function implicitly means
+    /// all readys collected before have been persisted.
+    pub fn on_persist_last_ready(&mut self) -> PersistLastReadyResult {
+        self.on_persist_ready(self.max_number);
```

Comment:
I think this may also assume that all previous `Ready` has been `commit_ready()` meaning that they all write to the file system (but not `fsync`). It may possible that the user gets a new `Ready` and calling `on_persist_last_ready`  before `commit_ready()` it

Reply:
Yes. But I think we should not worry about it. (I haven't added the doc for async ready yet)
I think maybe we can add some checking code for flow control in order to find users' wrong behavior.